code/modules/modular_computers/file_system/programs/engineering/power_monitor.dm

```diff
@@ -84,32 +84,94 @@
 	grid_sensors = list()
 	var/connected_z_levels = GetConnectedZlevels(get_host_z())
 	for(var/obj/machinery/power/sensor/S in SSmachines.machinery)
 		if(get_z(S) in connected_z_levels) // Consoles have range on their Z-Level. Sensors with long_range var will work between Z levels.
 			grid_sensors += S
-			GLOB.destroyed_event.register(S, src, /datum/nano_module/power_monitor/proc/remove_sensor)
+			GLOB.destroyed_event.register(S, src, /datum/nano_module/program/power_monitor/proc/remove_sensor)
 
-/datum/nano_module/power_monitor/proc/remove_sensor(var/removed_sensor, var/update_ui = TRUE)
+/datum/nano_module/program/power_monitor/proc/remove_sensor(var/removed_sensor, var/update_ui = TRUE)
 	if(active_sensor == removed_sensor)
 		active_sensor = null
 		if(update_ui)
 			SSnano.update_uis(src)
 	grid_sensors -= removed_sensor
-	GLOB.destroyed_event.unregister(removed_sensor, src, /datum/nano_module/power_monitor/proc/remove_sensor)
+	GLOB.destroyed_event.unregister(removed_sensor, src, /datum/nano_module/program/power_monitor/proc/remove_sensor)
+
+/datum/nano_module/program/power_monitor/proc/is_sysadmin(var/mob/user)
+	if(program)
+		return program.can_run(user, program.computer.get_network())
+	return FALSE
 
 // Allows us to process UI clicks, which are relayed in form of hrefs.
-/datum/nano_module/power_monitor/Topic(href, href_list)
+/datum/nano_module/program/power_monitor/Topic(href, href_list, state)
 	if(..())
 		return 1
+
+	var/mob/user = usr
+
 	if( href_list["clear"] )
 		active_sensor = null
 		. = 1
 	if( href_list["refresh"] )
 		refresh_sensors()
 		. = 1
+
+	if(href_list["toggle_breaker"])
+		var/obj/machinery/power/apc/A = locate(href_list["toggle_breaker"])
+
+		if(!CanInteract(user, state) || QDELETED(A))
+			return 0
+
+		A.operating = !A.operating
+
+	if(href_list["toggle_powerchannel_equip"] || href_list["toggle_powerchannel_light"] || href_list["toggle_powerchannel_enviro"]) //I'm sure there's a better way to do this.
+		var/obj/machinery/power/apc/A
+		var/powerchannel = 0
+		var/power_setting
+
+		if(href_list["toggle_powerchannel_equip"])
+			A = locate(href_list["toggle_powerchannel_equip"])
+			powerchannel = 1
+
+		if(href_list["toggle_powerchannel_light"])
+			A = locate(href_list["toggle_powerchannel_light"])
+			powerchannel = 2
+
+		if(href_list["toggle_powerchannel_enviro"])
+			A = locate(href_list["toggle_powerchannel_enviro"])
+			powerchannel = 3
+
+		if(A.is_critical && !is_sysadmin(user))
+			to_chat(user, SPAN_WARNING("Unsyndicated connections require system administrator access."))
+			return 0
+
+		var/list/answers_list = list("ON", "OFF", "AUTO")
+		var/input = input("Select a power channel mode.", "Power Channel Mode") as null|anything in answers_list
+
+		if(!CanInteract(user, state) || QDELETED(A))
+			return 0
+
+		switch(input)
+			if("ON")
+				power_setting = 3
+			if("OFF")
+				power_setting = 0
+			if("AUTO")
+				power_setting = 4
```

Comment:
There's a proc that lets you avoid some of the magic numbers. It may be better to make the whole thing into a proc.